pkg/netns/netns.go

```diff
@@ -4,88 +4,302 @@
 package netns
 
 import (
-	"errors"
 	"fmt"
 	"os"
-	"os/exec"
+	"path"
 	"path/filepath"
+	"runtime"
 	"strings"
+	"sync"
 
-	"github.com/containernetworking/plugins/pkg/ns"
-	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+
+	// TODO consider dropping this dep entirely - it is only used here now
+	"github.com/vishvananda/netns"
 )
 
-// RemoveIfFromNetNSIfExists removes the given interface from the given
-// network namespace.
+const bindMountPath = "/run/netns" /* Bind mount path for named netns */
+
+type netNSHandle struct {
+	handle int
+	Name   string
+}
+
+type NetNsHandle interface {
+	// Returns a file descriptor representing this object's network namespace
+	Fd() int
+
+	// Cleans up this instance of the network namespace; if this instance
+	// is the last user the namespace will be destroyed. Will Unpin first if
+	// instance is currently Pin-ed.
+	Close() error
+
+	// "Pins" or bind-mounts the netns created here to a specific path in nsfs.
+	// This ensures that the netns can have a lifetime beyond the creating process
+	// and this specific struct.
+	Pin() (string, error)
+
+	// "Unpins" or removes an existing bind-mount for the current network namespace handle in nsfs,
+	// if one exists. An unpinned netns will not have a lifetime beyond that of the creating process.
+	Unpin() error
+
+	// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+	//
+	// Note that since network namespaces are set *per thread*, both the current thread and the new thread will
+	// be locked with `runtime.LockOSThread()` during this operation.
+	//
+	// The provided closure should _not_ call runtime.UnlockOSThread() at all,
+	// or else undefined behavior should be expected.
+	Do(runInside func(NetNsHandle) error) error
```

Comment:
This may not be needed.

We have some existing code where we do things with e.g. interfaces within the Do closure and need a contract that the FD of the netns is available inside the closure.

For now in this WIP, I am just doing it this way so moving that code is straightforward.